integration/cli_test.go

```diff
@@ -316,88 +337,166 @@ func Test3LOLoopbackFlow(t *testing.T) {
 				"--disableAutoOpenConsentPage"},
 			"curl-3lo-loopback.golden",
 			false,
 		},
-		/*
-			{
-				"fetch; 3lo loopback cached",
-				[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-3lo-loopback.json", "--consentPageInteractionTimeout", "1", "--consentPageInteractionTimeoutUnits", "seconds"},
-				"fetch-3lo-cached.golden",
-				false,
-			},
-			{
-				"fetch; 3lo loopback insert expired token into cache",
-				[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json",
-				"--consentPageInteractionTimeout", "1", "--consentPageInteractionTimeoutUnits", "seconds"},
-				"fetch-3lo.golden",
-				false,
-			},
-			{
-				"fetch; 3lo loopback cached; token expired",
-				[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json",
-				"--consentPageInteractionTimeout", "1", "--consentPageInteractionTimeoutUnits", "seconds"},
-				"fetch-3lo.golden",
-				false,
-			},
-			{
-				"fetch; 3lo loopback cached; refresh expired token",
-				[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json", "--refresh",
-				"--consentPageInteractionTimeout", "1", "--consentPageInteractionTimeoutUnits", "seconds"},
-				"fetch-3lo-cached.golden",
-				false,
-			},*/
+		{
+			"fetch; 3lo loopback cached",
+			[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-3lo-loopback.json",
+				"--disableAutoOpenConsentPage",
+				"--consentPageInteractionTimeout", "30", "--consentPageInteractionTimeoutUnits", "seconds"},
+			"fetch-3lo-cached.golden",
+			false,
+		},
+		{
+			"fetch; 3lo loopback insert expired token into cache",
+			[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json",
+				"--disableAutoOpenConsentPage",
+				"--consentPageInteractionTimeout", "30", "--consentPageInteractionTimeoutUnits", "seconds"},
+			"fetch-3lo-loopback.golden",
+			false,
+		},
+		{
+			"fetch; 3lo loopback cached; token expired",
+			[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json",
+				"--disableAutoOpenConsentPage",
+				"--consentPageInteractionTimeout", "30", "--consentPageInteractionTimeoutUnits", "seconds"},
+			"fetch-3lo-loopback.golden",
+			false,
+		},
+		{
+			"fetch; 3lo loopback cached; refresh expired token",
+			[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json", "--refresh",
+				"--disableAutoOpenConsentPage",
+				"--consentPageInteractionTimeout", "30", "--consentPageInteractionTimeoutUnits", "seconds"},
+			"fetch-3lo-cached.golden",
+			false,
+		},
+	}
+
+	type LoopbackLogic struct {
```

Comment:
I think we need a high-level comment on why the loopback pre and post logic is needed, to explain port selection and loopback triggering etc. Also, the current logic has a process to "restore credentials file" - what if the test fails in the middle and the credentials file doesn't get restored? I feel like it may be better to create a new temp cred file ONCE at the start of the loopback tests and delete the file in a deferred block at the end of the loopback tests - so you don't have to look for an available port each test or do clean up of the file for each test - at that point, you may be able to just have a single logic step to trigger the loopback instead of pre/post. WDYT?

Reply:
Hi Andy, 

the following has been added:
- temp file is created and deleted per test case
- added logic to prevent looping forever when trying to reach for the code and state endpoint handler
- port is only searched for once and reused during each test case

the following was not changed:
- 2 step process: pre/post logic

NOTE: post helps on putting an end to the loopback logic in case the command execution ends unexpectedly/prematurely.

I believe this is a good compromise between your proposal and the original implementation. However, I am open to new suggestions.